src/ArduinoIoTCloudTCP.cpp

```diff
@@ -404,34 +406,58 @@ ArduinoIoTCloudTCP::State ArduinoIoTCloudTCP::handle_SubscribeMqttTopics()
     return State::ConnectPhy;
   }
 
-  if (!_mqttClient.subscribe(_dataTopicIn))
+  unsigned long const now = millis();
+  bool const is_subscribe_retry_delay_expired = (now - _last_subscribe_request_tick) > AIOT_CONFIG_SUBSCRIBE_RETRY_DELAY_ms;
+  bool const is_first_subscribe_request = (_last_subscribe_request_cnt == 0);
+  if (is_first_subscribe_request || is_subscribe_retry_delay_expired)
   {
-    DEBUG_ERROR("ArduinoIoTCloudTCP::%s could not subscribe to %s", __FUNCTION__, _dataTopicIn.c_str());
-#if !defined(__AVR__)
-    DEBUG_ERROR("Check your thing configuration, and press the reset button on your board.");
-#endif
-    return State::SubscribeMqttTopics;
-  }
+    if (_last_subscribe_request_cnt > AIOT_CONFIG_SUBSCRIBE_MAX_RETRY_CNT)
+    {
+      _last_subscribe_request_cnt = 0;
+      _last_subscribe_request_tick = 0;
+      _mqttClient.stop();
+      execCloudEventCallback(ArduinoIoTCloudEvent::DISCONNECT);
+      return State::ConnectPhy;
+    }
 
-  if (_shadowTopicIn != "")
-  {
-    if (!_mqttClient.subscribe(_shadowTopicIn))
+    _last_subscribe_request_tick = now;
+    _last_subscribe_request_cnt++;
+
+    if (!_mqttClient.subscribe(_dataTopicIn))
     {
-      DEBUG_ERROR("ArduinoIoTCloudTCP::%s could not subscribe to %s", __FUNCTION__, _shadowTopicIn.c_str());
+      DEBUG_ERROR("ArduinoIoTCloudTCP::%s could not subscribe to %s", __FUNCTION__, _dataTopicIn.c_str());
 #if !defined(__AVR__)
       DEBUG_ERROR("Check your thing configuration, and press the reset button on your board.");
 #endif
       return State::SubscribeMqttTopics;
     }
-  }
 
-  DEBUG_INFO("Connected to Arduino IoT Cloud");
-  execCloudEventCallback(ArduinoIoTCloudEvent::CONNECT);
+    if (_shadowTopicIn != "")
+    {
+      if (!_mqttClient.subscribe(_shadowTopicIn))
+      {
+        DEBUG_ERROR("ArduinoIoTCloudTCP::%s could not subscribe to %s", __FUNCTION__, _shadowTopicIn.c_str());
+#if !defined(__AVR__)
+        DEBUG_ERROR("Check your thing configuration, and press the reset button on your board.");
+#endif
+        return State::SubscribeMqttTopics;
+      }
+    }
 
-  if (_shadowTopicIn != "")
-    return State::RequestLastValues;
-  else
-    return State::Connected;
+    _last_subscribe_request_cnt = 0;
+    _last_subscribe_request_tick = 0;
+
+    DEBUG_INFO("Connected to Arduino IoT Cloud");
+    execCloudEventCallback(ArduinoIoTCloudEvent::CONNECT);
+
+    if (_shadowTopicIn != "")
+      return State::RequestLastValues;
+    else
+      return State::Connected;
+
+  }
+
+  return State::SubscribeMqttTopics;
```

Comment:
I personally would write it in the following way to make it more readable (to avoid nested if), but it's just an opinion:
```C++
  if (!is_first_subscribe_request && !is_subscribe_retry_delay_expired)
  {
    return State::SubscribeMqttTopics;
  }

  if (_last_subscribe_request_cnt > AIOT_CONFIG_SUBSCRIBE_MAX_RETRY_CNT)
  {
    _last_subscribe_request_cnt = 0;
    _last_subscribe_request_tick = 0;
    _mqttClient.stop();
    execCloudEventCallback(ArduinoIoTCloudEvent::DISCONNECT);
    return State::ConnectPhy;
  }

  if (!_mqttClient.subscribe(_dataTopicIn))
  {
    DEBUG_ERROR("ArduinoIoTCloudTCP::%s could not subscribe to %s", __FUNCTION__, _dataTopicIn.c_str());
#if !defined(__AVR__)
    DEBUG_ERROR("Check your thing configuration, and press the reset button on your board.");
#endif
    _last_subscribe_request_tick = now;
    _last_subscribe_request_cnt++;
    return State::SubscribeMqttTopics;
  }

  if (_shadowTopicIn != "")
  {
    if (!_mqttClient.subscribe(_shadowTopicIn))
    {
      DEBUG_ERROR("ArduinoIoTCloudTCP::%s could not subscribe to %s", __FUNCTION__, _shadowTopicIn.c_str());
#if !defined(__AVR__)
      DEBUG_ERROR("Check your thing configuration, and press the reset button on your board.");
#endif
      _last_subscribe_request_tick = now;
      _last_subscribe_request_cnt++;
      return State::SubscribeMqttTopics;
    }
  }

  DEBUG_INFO("Connected to Arduino IoT Cloud");
  execCloudEventCallback(ArduinoIoTCloudEvent::CONNECT);
  _last_subscribe_request_cnt = 0;
  _last_subscribe_request_tick = 0;
  if (_shadowTopicIn != "")
  {
    return State::RequestLastValues;
  }
  return State::Connected;
```

Reply:
I second @polldo's comment.

Reply:
Thanks @polldo:
 - nested if: i would leave this one as it is 
```
   if (_shadowTopicIn != "")
   {
     if (!_mqttClient.subscribe(_shadowTopicIn))
```
since it wi be removed when we will merge the thing_id_discovery PR #297 
 - which is the advantage moving this part inside the ifs? Personally i would leave it outside.
```
 _last_subscribe_request_tick = now;
 _last_subscribe_request_cnt++;
```

Reply:
>Personally i would leave it outside.

do it! 
I moved them to make it clear that those statements depended on the subscription fail